crates/accelerate/src/high_level_synthesis.rs

```diff
@@ -0,0 +1,252 @@
+}
+
+#[pymethods]
+impl QubitTracker {
+    #[new]
+    pub fn new(num_qubits: usize) -> Self {
+        QubitTracker {
+            num_qubits,
+            state: vec![false; num_qubits],
+            enabled: vec![true; num_qubits],
+            ignored: vec![false; num_qubits],
+        }
+    }
+
+    /// Sets state of the given qubits to dirty
+    fn set_dirty(&mut self, qubits: Vec<usize>) {
+        for q in qubits {
+            self.state[q] = false;
+        }
+    }
+
+    /// Sets state of the given qubits to clean
+    fn set_clean(&mut self, qubits: Vec<usize>) {
+        for q in qubits {
+            self.state[q] = true;
+        }
+    }
+
+    /// Disables using the given qubits
+    fn disable(&mut self, qubits: Vec<usize>) {
+        for q in qubits {
+            self.enabled[q] = false;
+        }
+    }
+
+    /// Enable using the given qubits
+    fn enable(&mut self, qubits: Vec<usize>) {
+        for q in qubits {
+            self.enabled[q] = true;
+        }
+    }
+
+    /// Returns the number of enabled clean qubits, ignoring the given qubits
+    /// ToDo: check if it's faster to avoid using ignored
+    fn num_clean(&mut self, ignored_qubits: Vec<usize>) -> usize {
+        for q in &ignored_qubits {
+            self.ignored[*q] = true;
+        }
+
+        let count = (0..self.num_qubits)
+            .filter(|q| !self.ignored[*q] && self.enabled[*q] && self.state[*q])
+            .count();
+
+        for q in &ignored_qubits {
+            self.ignored[*q] = false;
+        }
+
+        count
+    }
+
+    /// Returns the number of enabled dirty qubits, ignoring the given qubits
+    /// ToDo: check if it's faster to avoid using ignored
+    fn num_dirty(&mut self, ignored_qubits: Vec<usize>) -> usize {
+        for q in &ignored_qubits {
+            self.ignored[*q] = true;
+        }
+
+        let count = (0..self.num_qubits)
+            .filter(|q| !self.ignored[*q] && self.enabled[*q] && !self.state[*q])
+            .count();
+
+        for q in &ignored_qubits {
+            self.ignored[*q] = false;
+        }
+
+        count
+    }
+
+    /// Get `num_qubits` enabled qubits, excluding `ignored_qubits`, and returning the
+    /// clean qubits first.
+    /// ToDo: check if it's faster to avoid using ignored
+    fn borrow(&mut self, num_qubits: usize, ignored_qubits: Vec<usize>) -> Vec<usize> {
+        for q in &ignored_qubits {
+            self.ignored[*q] = true;
+        }
+
+        let clean_ancillas = (0..self.num_qubits)
+            .filter(|q| !self.ignored[*q] && self.enabled[*q] && self.state[*q]);
+        let dirty_ancillas = (0..self.num_qubits)
+            .filter(|q| !self.ignored[*q] && self.enabled[*q] && !self.state[*q]);
+        let out: Vec<usize> = clean_ancillas
+            .chain(dirty_ancillas)
+            .take(num_qubits)
+            .collect();
+
+        for q in &ignored_qubits {
+            self.ignored[*q] = false;
+        }
+        out
+    }
```

Comment:
Here's how it would look with the Hashset:
```suggestion
    /// Get `num_qubits` enabled qubits, excluding `ignored_qubits`, and returning the
    /// clean qubits first.
    /// ToDo: check if it's faster to avoid using ignored
    fn borrow(&mut self, num_qubits: usize, ignored_qubits: HashSet<usize>) -> Vec<usize> {
        let clean_ancillas = (0..self.num_qubits)
            .filter(|q| !ignored_qubits.contains(q) && self.enabled[*q] && self.state[*q]);
        let dirty_ancillas = (0..self.num_qubits)
            .filter(|q| !self.ignored[*q] && self.enabled[*q] && !self.state[*q]);
        clean_ancillas
            .chain(dirty_ancillas)
            .take(num_qubits)
            .collect()
    }
```